micro_http/src/server.rs

```diff
@@ -0,0 +1,681 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+extern crate epoll;
+
+use std::io::{Error, Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::path::Path;
+
+use common::Version;
+pub use common::{ConnectionError, RequestError};
+use connection::HttpConnection;
+use request::Request;
+use response::{Response, StatusCode};
+use std::collections::HashMap;
+
+static SERVER_FULL_ERROR_MESSAGE: &'static [u8] = b"HTTP/1.1 500\r\n\
+                                            Server: Firecracker API\r\n\
+                                            Connection: close\r\n\
+                                            Content-Length: 14\r\n\r\nServer is full";
+const MAX_CONNECTIONS: usize = 10;
+
+type Result<T> = std::result::Result<T, ServerError>;
+
+/// Errors pertaining to the server.
+#[derive(Debug)]
+pub enum ServerError {
+    /// Epoll operations failed.
+    IOError(Error),
+    /// Error from one of the connections.
+    ConnectionError(ConnectionError),
+    /// Server maximum capacity has been reached.
+    ServerFull,
+}
+
+/// Wrapper over `Request` which adds an identification token.
+pub struct RequestWithID {
```

Comment:
`RequestWithID` was implemented to add multiplexing logic to a request. The ID should not be a part of a `Request` object as it is not relevant to the request itself, but we wanted to provide a transport layer within the code for a request.

We also wanted to mask the ID because it is not relevant for the user of the `HttpServer`. The alternative is changing the return type of `handle_notifications()` to `Vec<(Token, Request)>`.

If something relevant to a request might be added, we should do so in `request.rs`, in the `Request` structure, not in this one. This is a glorified wrapper.